### ISSUE TITLE
Victory and defeat conditions

### ISSUE THREAD

**Issue by sergiofv93** [jmespadero/cityGen#20]
This Issue is about the implementation of conditions for the Victory or the Defeat in the game.

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I have some questions.

First of all, I've been working in the implementation that we decided, I mean, to separate the functionality of vampires in another internal text file inside the cg-monsters.blend file.

However, I've considered that will be much better to separate those functionalities in blender files, not only in internal files. I mean, I think would be useful to have several blend files like 'cg-monsters.blend', 'cg-vampires.blend' and 'cg-winorlose.blend' files. That's because the Init() method that each internal python file needs has to be called from an object, in the cg-monsters.blend file case, represented by a invisible cube named 'AI_Manager'. This cube woulds need to add a new logic controller calling the Init method of the vampire_manager.py file, but it already has a Init controller for the AI_Manager.py internal file.

So I think it would be useful to have another blender file with another "Manager_Cube" inside which will call the Init() method of the Vampires_Manager.py internal file. And the same for a 'winorlose.blend' file. What do you think about it?

And now, the code question: I don't know how the 'cg-monsters.blend' library is loaded in the cityGen3D.py file. I mean, I know that the corresponding code are the next:
![image](https://user-images.githubusercontent.com/22729865/48448915-d179bc80-e7a0-11e8-8876-6f81e2f33efc.png)

But I don't know how is it possible, if the arg 'inputMonsterLibrary' does not exists, and the previous 'if' statement creates it empty. I need to know it, because if I want to implement what I have explained to you previously, I need to load the new 'cg-vampires.blend' file into the 'cityGen3D.py' file to include it in the output file and run it.

Thanks in advance,
greetings.

**Comment by jmespadero** [comment]
> First of all, I've been working in the implementation that we decided, I mean, to separate the functionality of vampires in another internal text file inside the cg-monsters.blend file.

Right, my idea was not a separate file inside of cg-monsters.blend, but a separate cg-vampires.blend file. 
This is a .blend file with the model of the vampire and a module to "populate" the streets of a city with vampires.  So, in the script cityGen3D.py you are free to include a player (the boy, the scarecrow or none), include a skyDome (night, day, or none), include monsters (or not) and include `dynamic decoration, as vampires, or not. This way, you may choose to create a city with sunny day and butterflies, or a doomed city with Monters, vampires and night. Or any other combination...

>So I think it would be useful to have another blender file with another "Manager_Cube" inside which >will call the Init() method of the Vampires_Manager.py internal file. And the same for a >'winorlose.blend' file. What do you think about it?

Yes, that's the idea...

>And now, the code question: I don't know how the 'cg-monsters.blend' library is loaded in the cityGen3D.py file. 

`inputMonsterLibrary` is defined in the file cg-config.json. Its current value is

```json
"inputMonsterLibrary": ["cg-monsters.blend"],
```
So, just change to 
```json
"inputMonsterLibrary": ["cg-monsters.blend", "cg-vampires.blend"],
```

and you shoud import both files.

**Comment by jmespadero** [comment]
And now something more about the cube called `AI_Manager`...

This object is currently being used with three purposes:

1. To call the Init() method in AI_Manager module, by a connection with an "always sensor"

2. To hold the keyboard sensors to keys "1", "2", "3", "4", used in the separate monsters demo.
     This is not really needed for the game.

3. To be a hook where citygen3D.py can write information to be read by AI_manager module.
     This mean that citygen3D.py will search  `AI_Manager` and inject some data there:
      * A copy of 'cg-data.json' file, inside of a fake controller named 'cg-data.json'
      * A copy of 'cg-ia.json' file, inside of a fake controller named 'cg-ia.json'
      * An object property named 'iniMonsters' with the index of the starting point for monsters.

      Citygen3D need to put that info somewhere so AI_Manager can found the info when executed. This is done inside of AI_Manager.Init() with this code:
    ```python
    # Try to read from cg-data.json controller
    if 'cg-data.json' in own.controllers:
        print("Parsing file controllers['cg-data.json'].script")
        data = json.loads(own.controllers['cg-data.json'].script)
    ```

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I have find a problem about knowing the points that belong to the wall in the "vertices" city.data.json file. I would compare the vertices list with the wallVertices list, but I see that the same point  has different coordinates. For example, I can see a point (7, 207), and I see that the most similar point in the wallVertices list is (7, 203), so I can't know if a wallVertices point is contained by the vertices list.

![seleccion_007](https://user-images.githubusercontent.com/22729865/50233290-66c92b80-03b3-11e9-9b26-6ce26c2ec80c.png)


Thanks in advance,
Kind Regards.

**Comment by jmespadero** [comment]
As you can see in the city.map.verbose.svg, the wallVertices **are not part** of city corners.

![city map verbose](https://user-images.githubusercontent.com/6451497/50242542-faf2bd00-03ca-11e9-9a9e-3a78fe41c17f.png)

The towers are conveniently displaced, so every coordinate in "vertices" can be considered internal to the city and player and monsters can walk near of the wall.

If you really need a list with the indexes of vertices not in the external streets that goes near the wall, you can  _negate_ the "externalPoints" list

```python
        internalPoints = [i for i in range(len(vertices)) if i not in externalPoints]
```

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

thanks for your help. However, I get a fail trying to execute the "random.sample(internalPoints, X)" about the recursion limit. I've been taking a look on internet ([https://stackoverflow.com/questions/8177073/python-maximum-recursion-depth-exceeded](https://stackoverflow.com/questions/8177073/python-maximum-recursion-depth-exceeded)) and some users recomend me to use the next code:
```
import sys
sys.setrecursionlimit(10000) 
```

but it does not work at all. I'm getting this error:
![error](https://user-images.githubusercontent.com/22729865/50276797-1d282180-0443-11e9-89eb-ebefb50e7d8f.jpg)

Same problem for random.sample and random.shuffle methods. Can you suggest me any possible solution?

Thanks in advance,
kind regards.

**Comment by jmespadero** [comment]
My calls to `random.sample(internalPoints, X)` works as expected.  Why are you using recursion here?

**Comment by sergiofv93** [comment]
Mmmmm,

I'm calling to `random.sample(list, x)` with python3 in my console and works propperly... Only blender file algorithm is failing

I'll keep on investigating and will post here any founded solution.

Thanks,
Kind Regards.

**Comment by jmespadero** [comment]
Can you paste the output of 'print(internalPoints)' ?

**Comment by sergiofv93** [comment]
Hi @jmespadero 

now the `random.sample` method is working properly. The released pumpkins number is now loaded from a game property named "pumpkins" which is initialized as an integer into the AI_Manager object.

Whats more, the Victory condition is implemented. The functionality puts over the current scene another scene with the VICTORY text. After showing this text, the game will be finished.

![imagen](https://user-images.githubusercontent.com/22729865/51241713-a1a48c00-197e-11e9-8564-c08a49734886.png)

Thanks in advance,
greetings.

**Comment by jmespadero** [comment]
Why AI_Manager? Should not be WL_Manager?

**Comment by sergiofv93** [comment]
Ups, you are right. I'll fix it.

Thanks!

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I've pushed new content to the repository. In this case, I've implemented the algorithm that provides the behaviour of the compass and the life_bar. The life bar works propperly when a spider strikes the player, and I have been working with mathutils.Euler objects to give the correct angle to the compass.

But I still have a problem. I have been looking for a solution for giving a rotation order to the euler, and I found that we can asign the order in the class constructor, just like this:

`compass_wise.localOrientation = mathutils.Euler(angle, 'ZXY')`

But it does not work propperly. It seems like the code ignores the given order and applies the X rotation first. What could the problem be¿?

Thanks in advance,
greetings

**Comment by jmespadero** [comment]
You must provide 3 angles, not one.
Also, do an explicit conversion to matrix when assigning to localOrientation. It is supposed that blender do that, but some people has reported problems when assigning directly the Euler rotation.

```py
compass_wise.localOrientation = mathutils.Euler((angleX, angleY, angleZ), 'ZXY').to_matrix()
```

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I have finished the implementation of the HUD and the on-lose conditions algorithm, so we could close this issue. This issue has been too long because I have implemented both win and lose functionalities.

In this issue, the changes made about winning the game were::
1º.- Adding the 'pumpkin' animated model to the repository.
2º.- Releasing of pumpkins into the city.
3º.- New algorithm: when the player picks up a specific number of pumpkins, the gates will open.
4º.- Victory splash screen is deployed when the player pass through the opened gates.

What's more, the next contents about losing the game were added too:
5º.-  Added a 'life bar' in the top left screen edge.
6.º.- Added a 'compass' in the bottom right screen edge.
7º.- New algorithm: life bar decreases when a monster touches the player.
8º.- New algorithm: compass is always marking the direction of the north.
9º.- New algorithm: when the life_bar is empty, the player looses the game.

This new functionalities will be well explained in the final documentation of the proyect, but I will write a small summary down below.

The pumpkin and compass models have been downloaded from the [https://www.blendswap.com/](https://www.blendswap.com/) web. The pumpkins algorithm is in charge of computing how many pumpkins has the player picked. When a specific number of pumpkins (defined by a property) have been collected, the main gate of the wall will be open and the player will can pass through the gates and win the game (a VICTORY screen will be displayed ).
The pumpkins are released only in the intersections of the streets:

![yellowBallsDebbuging](https://user-images.githubusercontent.com/22729865/54882304-dd654180-4e58-11e9-87b4-56cc485899b5.png)


What's more, the player could lose the game. The initial idea for this purpose was to detect collisions between the monsters and the player. This collisions decrease a counter and the algorithm is in charge of evaluating that counter and, if it reaches the 0 value, the game must end.

What's more, a HUD screen has been implemented. This screen contains a life bar and a compass. the life bar shows the level of life which owns the player, like the most of the videogames. The compass is an independent object which has nothing related to the on-lose algorithm. It's only working for indicating the direction of the north to the player.

![lifebar](https://user-images.githubusercontent.com/22729865/54882285-a7c05880-4e58-11e9-8133-3656761b7afe.png)


The behaviour is quite simple. The player begins the game looking to the north, always. Knowing it, the algorithm only gets the rotation of the player in each logic blender tic and computed the position of the compass wise.

![compass](https://user-images.githubusercontent.com/22729865/54882299-c6265400-4e58-11e9-9df6-7366865740e4.png)

Thanks for your atemption,
greetings.

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

with the recently merged PullRequest#22, I think we are able to close this issue.
What's more, I need you to create a new "MainMenu" issue to post the new questions about the development.

Thanks in advance,
kind regards.

**Comment by jmespadero** [comment]
As you want, but I though you were to add a background to the compass instead of leaving white.
